forest/benchmarking/superoperator_tools.py

```diff
@@ -363,3 +367,103 @@ def computational2pauli_basis_matrix(dim) -> np.ndarray:
+    r"""
+    Implements a Pauli twirl of a chi matrix (aka a process matrix).
+
+    See the folloiwng reference for more details
+
+    [SPICC] Scalable protocol for identification of correctable codes
+            Silva et al.,
+            PRA 78, 012347 2008
+            http://dx.doi.org/10.1103/PhysRevA.78.012347
+            https://arxiv.org/abs/0710.1900
+
+    Note: Pauli twirling a quantum channel can give rise to a channel that is less noisy; use with
+    care.
+
+    :param chi_matrix:  a dim**2 by dim**2 chi or process matrix
+    :return: dim**2 by dim**2 chi or process matrix
+    """
+    return np.diag(chi_matrix.diagonal())
+
+
+#TODO Honest approximations for Channels that act on one or MORE qubits.
+
+# ==================================================================================================
+# Apply channel
+# ==================================================================================================
+def apply_kraus_ops_2_state(kraus_ops: Sequence[np.ndarray], state: np.ndarray) -> np.ndarray:
+    r"""
+    Apply a quantum channel, specified by Kraus operators, to state.
+
+    The Kraus operators need not be square.
+
+    :param kraus_ops: A list or tuple of N Kraus operators, each operator is M by dim ndarray
+    :param state: A dim by dim ndarray which is the density matrix for the state
+    :return: M by M ndarray which is the density matrix for the state after the action of kraus_ops
+    """
+    if isinstance(kraus_ops, np.ndarray):  # handle input of single kraus op
+        if len(kraus_ops[0].shape) < 2:  # first elem is not a matrix
+            kraus_ops = [kraus_ops]
+
+    dim, _ = state.shape
+    rows, cols = kraus_ops[0].shape
+
+    if dim != cols:
+        raise ValueError("Dimensions of state and Kraus operator are incompatible")
+
+    new_state = np.zeros((rows, rows))
+    for M in kraus_ops:
+        new_state += M @ state @ np.transpose(M.conj())
+
+    return new_state
+
+
+def apply_choi_matrix_2_state(choi: np.ndarray, state: np.ndarray) -> np.ndarray:
+    r"""
+    Apply a quantum channel, specified by a Choi matrix (using the column stacking convention),
+    to a state.
+
+    The Choi matrix is a dim**2 by dim**2 matrix and the state rho is a dim by dim matrix. The
+    output state is
+
+    rho_{out} = Tr_{A}[(rho^T \otimes Id) Choi_matrix ],
+
+    where Tr_{A} is the partial trace over hilbert space H_A with respect to the tensor product
+    H_A \otimes H_B, and T denotes transposition.
+
+
+    :param choi: a dim**2 by dim**2 matrix
+    :param state: A dim by dim ndarray which is the density matrix for the state
+    :return: a dim by dim matrix.
+    """
+    dim = int(np.sqrt(np.asarray(choi).shape[0]))
+    dims = [dim, dim]
+    tot_matrix = np.kron(state.transpose(), np.identity(dim)) @ choi
+    return partial_trace(tot_matrix, [1], dims)
+
+
+def apply_lioville_matrix_2_state(superoperator: np.ndarray, state: np.ndarray) -> np.ndarray:
```

Comment:
I don't think we should include this method. It is very straightforward to do the matrix multiplication and numpy already does the dimension check with a clear error message. 

Further, the current implementation takes in a vectorized state whereas apply_choi_matrix_2_state takes in a density matrix. I think this ambiguity alone would make it easier for me to do this myself than to call a method where I have to guess what form the state input should take.

Reply:
done!